pkg/bundle/controller.go

```diff
@@ -103,28 +93,52 @@ func AddBundleController(ctx context.Context, mgr manager.Manager, opts Options)
 	}
 
+	// This informer setup allow us to use the informers from the auxiliary,
+	// namespace-scoped cache to trigger event handlers of the bundle
+	// controller.
+	// We sync Secrets to the auxiliary, namespace scoped cache rather than
+	// the cache that gets created with the c/r manager. This is a temporary
+	// fix to allow us to scope down RBAC for Secrets. See
+	// https://github.com/kubernetes-sigs/controller-runtime/pull/2261#discussion_r1211640590
+	// for context.
+	// For convenience, we also sync ConfigMaps, Namespaces and Bundles to
+	// the same cache as that allows us to use a single cache client to GET
+	// all these cached resources inside reconcile loop.
+
+	configMapInformer, err := sourceCache.GetInformer(ctx, &corev1.ConfigMap{})
+	if err != nil {
+		return fmt.Errorf("error creating ConfigMaps informer from namespace-scoped cache: %w", err)
+	}
+	secretInformer, err := sourceCache.GetInformer(ctx, &corev1.Secret{})
+	if err != nil {
+		return fmt.Errorf("error creating Secrets informer from namespace-scoped cache: %w", err)
+	}
+	bundleInformer, err := sourceCache.GetInformer(ctx, &trustapi.Bundle{})
+	if err != nil {
+		return fmt.Errorf("error creating Bundle informer from namespace-scoped cache: %w", err)
+	}
+	namespaceInformer, err := sourceCache.GetInformer(ctx, &corev1.Namespace{})
+	if err != nil {
+		return fmt.Errorf("error creating Bundle informer from namespace-scoped cache: %w", err)
+	}
+
 	// Only reconcile config maps that match the well known name
 	if err := ctrl.NewControllerManagedBy(mgr).
 		Named("bundles").
 
 		////// Targets //////
 
-		// Reconcile over owned ConfigMaps in all Namespaces. Only cache metadata.
-		// These ConfigMaps will be Bundle Targets
-		Watches(&source.Kind{Type: new(corev1.ConfigMap)}, &handler.EnqueueRequestForOwner{
-			OwnerType:    new(trustapi.Bundle),
-			IsController: true,
-		}, builder.OnlyMetadata).
-
-		////// Sources //////
+		// Reconcile a Bundle on events against a ConfigMap that it
+		// owns. Only cache ConfigMap metadata.
+		WatchesMetadata(&corev1.ConfigMap{}, handler.EnqueueRequestForOwner(mgr.GetScheme(), mgr.GetRESTMapper(), &trustapi.Bundle{}, handler.OnlyControllerOwner())).
 
 		// Reconcile trust.cert-manager.io Bundles
-		Watches(source.NewKindWithCache(new(trustapi.Bundle), sourceCache), &handler.EnqueueRequestForObject{}).
+		WatchesRawSource(&source.Informer{Informer: bundleInformer}, &handler.EnqueueRequestForObject{}).
 
 		// Watch all Namespaces. Cache whole Namespaces to include Phase Status.
 		// Reconcile all Bundles on a Namespace change.
-		Watches(source.NewKindWithCache(new(corev1.Namespace), sourceCache), handler.EnqueueRequestsFromMapFunc(
-			func(obj client.Object) []reconcile.Request {
+		WatchesRawSource(&source.Informer{Informer: namespaceInformer}, handler.EnqueueRequestsFromMapFunc(
+			func(_ context.Context, obj client.Object) []reconcile.Request {
```

Comment:
suggestion: This function also uses context by the looks of it!

```suggestion
			func(ctx context.Context, obj client.Object) []reconcile.Request {
```

Reply:
Thanks for spotting those!